pkg/utils/resources/resources.go

```diff
@@ -93,21 +95,78 @@ func Subtract(lhs, rhs v1.ResourceList) v1.ResourceList {
 	return result
 }
 
-// Ceiling calculates the max between the sum of container resources and max of initContainers
-func Ceiling(pod *v1.Pod) v1.ResourceRequirements {
-	var resources v1.ResourceRequirements
+// inspired from https://github.com/kubernetes/kubernetes/blob/e2afa175e4077d767745246662170acd86affeaf/pkg/api/v1/resource/helpers.go#L96
+// https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/
+// Ceiling calculates the max between the sum of container resources and max of initContainers along with sidecar feature consideration
+func podRequests(pod *v1.Pod) v1.ResourceList {
+	requests := v1.ResourceList{}
+	restartableInitContainerReqs := v1.ResourceList{}
+	maxInitContainerReqs := v1.ResourceList{}
+
 	for _, container := range pod.Spec.Containers {
-		resources.Requests = MergeInto(resources.Requests, MergeResourceLimitsIntoRequests(container))
-		resources.Limits = MergeInto(resources.Limits, container.Resources.Limits)
+		MergeInto(requests, MergeResourceLimitsIntoRequests(container))
 	}
+
 	for _, container := range pod.Spec.InitContainers {
-		resources.Requests = MaxResources(resources.Requests, MergeResourceLimitsIntoRequests(container))
-		resources.Limits = MaxResources(resources.Limits, container.Resources.Limits)
+		containerReqs := MergeResourceLimitsIntoRequests(container)
+		if lo.FromPtr(container.RestartPolicy) == v1.ContainerRestartPolicyAlways {
+			MergeInto(requests, containerReqs)
+			MergeInto(restartableInitContainerReqs, containerReqs)
+			maxInitContainerReqs = MaxResources(maxInitContainerReqs, restartableInitContainerReqs)
+
+		} else {
+			maxInitContainerReqs = MaxResources(maxInitContainerReqs, Merge(containerReqs, restartableInitContainerReqs))
```

Comment:
Consider adding a comment around what's happening here for clarity. e.g. "Else, check whether the current container's resource requests combined with the restartableInitContainer requests are greater than the current max"